client/fingerprint/dynamic_host_volumes_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package fingerprint
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	"github.com/hashicorp/nomad/client/config"
+	hvm "github.com/hashicorp/nomad/client/hostvolumemanager"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/shoenig/test/must"
+)
+
+// this is more of a full integration test of:
+// fingerprint <- find plugins <- find executables
+func TestPluginsHostVolumeFingerprint(t *testing.T) {
+	cfg := &config.Config{DynamicHostVolumePluginPath: ""}
+	node := &structs.Node{Attributes: map[string]string{}}
+	req := &FingerprintRequest{Config: cfg, Node: node}
+	fp := NewPluginsHostVolumeFingerprint(testlog.HCLogger(t))
+
+	// this fingerprint is not mandatory, so no error should be returned
+	for name, path := range map[string]string{
+		"empty":        "",
+		"non-existent": "/nowhere",
+		"impossible":   "dynamic_host_volumes_test.go",
+	} {
+		t.Run(name, func(t *testing.T) {
+			resp := FingerprintResponse{}
+			cfg.DynamicHostVolumePluginPath = path
+			err := fp.Fingerprint(req, &resp)
+			must.NoError(t, err)
+			must.True(t, resp.Detected) // always true due to "mkdir" built-in
+		})
+	}
+
+	if runtime.GOOS == "windows" {
+		t.Skip("test scripts not built for windows") // db TODO(1.10.0)
+	}
+
+	// happy path: dir exists. this one will contain a single valid plugin.
+	tmp := t.TempDir()
+	cfg.DynamicHostVolumePluginPath = tmp
+
+	files := []struct {
+		name     string
+		contents string
+		perm     os.FileMode
+	}{
+		// only this first one should be detected as a valid plugin
+		{"happy-plugin", "#!/usr/bin/env sh\necho '0.0.1'", 0700},
+		{"not-a-plugin", "#!/usr/bin/env sh\necho 'not-a-version'", 0700},
+		{"unhappy-plugin", "#!/usr/bin/env sh\necho '0.0.2'; exit 1", 0700},
+		{"not-executable", "hello", 0400},
+	}
+	for _, f := range files {
+		must.NoError(t, os.WriteFile(filepath.Join(tmp, f.name), []byte(f.contents), f.perm))
+	}
+	// directories should be ignored
+	must.NoError(t, os.Mkdir(filepath.Join(tmp, "a-directory"), 0700))
+
+	// do the fingerprint
+	resp := FingerprintResponse{}
+	err := fp.Fingerprint(req, &resp)
+	must.NoError(t, err)
+	must.Eq(t, map[string]string{
+		"plugins.dhv.version.happy-plugin": "0.0.1",
+		"plugins.dhv.version.mkdir":        hvm.HostVolumePluginMkdirVersion, // built-in
+	}, resp.Attributes)
+
+	// do it again after deleting our one good plugin.
+	// repeat runs should wipe attributes, so nothing should remain.
+	node.Attributes = resp.Attributes
+	must.NoError(t, os.Remove(filepath.Join(tmp, "happy-plugin")))
+
+	resp = FingerprintResponse{}
+	err = fp.Fingerprint(req, &resp)
+	must.NoError(t, err)
+	must.Eq(t, map[string]string{
+		"plugins.dhv.version.happy-plugin": "", // will get cleaned up later
```

Comment:
What's this comment referring to?

Reply:
it refers to this docstring

https://github.com/hashicorp/nomad/blob/v1.9.3/client/fingerprint/structs.go#L52-L54
```go
// RemoveAttribute sets the given attribute to empty, which will later remove
// it entirely from the node
func (f *FingerprintResponse) RemoveAttribute(name string) {
```

just to clarify that the map key still existing is expected. my intuition was that the attribute should be _removed_, not just set to empty.

Reply:
Oh, huh. Yeah mine too. :grin: